### PR TITLE
#4307 - add indexes

### PR DESCRIPTION
**Pull request by BarnaBartha**

### CODE REVIEW

File: sormas-backend/src/main/resources/sql/sormas_schema.sql [hunk 1]
```diff
@@ -6467,4 +6467,44 @@ ALTER TABLE cases_history ADD COLUMN caseidentificationsource character varying(
 
 INSERT INTO schema_version (version_number, comment) VALUES (324, 'Case identification source #3420');
 
+-- 2020-02-09 Add indexes #4307
+CREATE INDEX IF NOT EXISTS idx_cases_epid_number ON cases USING gist (epidnumber gist_trgm_ops);
+CREATE INDEX IF NOT EXISTS idx_cases_person_id ON cases (person_id);
+CREATE INDEX IF NOT EXISTS idx_cases_region_id ON cases (region_id);
+CREATE INDEX IF NOT EXISTS idx_cases_district_id ON cases (district_id);
+
+CREATE INDEX IF NOT EXISTS idx_contact_person_id ON contact (person_id);
```
Comment (DavidBaldsiefen) [inline sormas-backend/src/main/resources/sql/sormas_schema.sql:6476]: Remove this line, index already exists

File: sormas-backend/src/main/resources/sql/sormas_schema.sql [hunk 2]
```diff
@@ -6467,4 +6467,44 @@ ALTER TABLE cases_history ADD COLUMN caseidentificationsource character varying(
...
+
+CREATE INDEX IF NOT EXISTS idx_samples_contact_id ON samples (associatedcontact_id);
+CREATE INDEX IF NOT EXISTS idx_samples_case_id ON samples (associatedcase_id);
+CREATE INDEX IF NOT EXISTS idx_samples_eventparticipant_id ON samples (associatedeventparticipant_id);
+
+CREATE INDEX IF NOT EXISTS idx_task_contact_id ON task (contact_id);
+CREATE INDEX IF NOT EXISTS idx_task_case_id ON task (caze_id);
+CREATE INDEX IF NOT EXISTS idx_task_event_id ON task (event_id);
+
+CREATE INDEX IF NOT EXISTS idx_visit_case_id ON visit (caze_id);
```
Comment (DavidBaldsiefen) [inline sormas-backend/src/main/resources/sql/sormas_schema.sql:6489]: Remove, already exists

**Review by DavidBaldsiefen — CHANGES_REQUESTED** [review]
Looks good to me. We will have to see how this performs on life systems. Also, these changes might lead to quite long transition times when servers with large databases are updated, as all these indexes need to be created first.

The new indexes in the contact and case directory could actually _slow down_ the case/contact import. So I am not sure if it is wise to implement them now. Maybe @MartinWahnschaffe can take a quick look?

**Comment by MartinWahnschaffe** [comment]
@DavidBaldsiefen The indices Barna added are all for relations that are at the heart of the system and are relevant for almost all queries done. I'm pretty sure this will outweigh the cost of maintaining the index.

@BarnaBartha I think we should also consider the references within the event participant. Also what do you think about adding indices to the reporting user of at least case and contact? The createUserFilter logic is executed for almost all queries and relies on this.

**Comment by BarnaBartha** [comment]
@MartinWahnschaffe added the indexes to eventParticipants>
The reporting user indices have already been added by Salva
https://github.com/hzi-braunschweig/SORMAS-Project/pull/4333/files

**Comment by MartinWahnschaffe** [comment]
@BarnaBartha Great. Please go ahead and merge once the checks have passed. Please provide the SQL for all indices with a short message to @StefanKock for the release notes page.

**Pull request merged**